src/mock.js

```diff
@@ -82,6 +82,21 @@ function addMockFunctionsToSchema({ schema, mocks = {}, preserveResolvers = fals
     return customMock;
   }
 
+  function assignResolveType(type) {
+    const fieldType = getNullableType(type);
+    const namedFieldType = getNamedType(fieldType);
+
+    if (namedFieldType instanceof GraphQLUnionType ||
+        namedFieldType instanceof GraphQLInterfaceType
+    ) {
+      const oldResolveType = namedFieldType.resolveType;
+      // the default `resolveType` always returns null. We add a fallback
+      // resolution that works with how unions and interface are mocked
+      namedFieldType.resolveType = (data, context, info) =>
+        oldResolveType(data, context, info) || info.schema.getType(data.typename);
```

Comment:
@itajaja If I understand correctly, `resolveType` only returns null if the schema doesn't define a `resolveType` function. If the user provided one however, that will be used, so it won't return `null`.

The "problem" actually comes from here: https://github.com/graphql/graphql-js/blob/516dfb9674b4dbde1ee17f715e4c0b6bc24a9f92/src/utilities/buildASTSchema.js#L393

Since that code uses an arrow function with no arguments(`() => null`), so one hacky way to figure out if there was no resolve function defined is to check `resolveType.length`, which will be 0 for the arrow function, but non-zero for any user-defined (and working) `resolveType`.

So in short:
* If `preserveResolvers` is `false`, overwrite `resolveType` with the mock `resolveType` function, no matter what.
* If `preserveResolvers` is `true`, check `resolveType.length`. If it's zero, overwrite `resolveType` with the mock, otherwise leave it alone.

Does that make sense? It's a bit hacky, but I think it should work, and I don't currently see any other way if we want to keep using `buildASTSchema`.